src/properties/mod.rs

```diff
@@ -143,13 +165,89 @@ macro_rules! impl_read_header {
 pub(crate) use impl_read;
 pub(crate) use impl_read_header;
 
+macro_rules! impl_write {
+    (options, $property:ident $(, $header_property:tt)*) => {
+        impl PropertyTrait for $property {
+            #[inline]
+            fn write<W: Write>(
+                &self,
+                cursor: &mut W,
+                include_header: bool,
+                options: &mut PropertyOptions,
+            ) -> Result<(), Error> {
+                if !include_header {
+                    return self.write_body(cursor, options);
+                }
+
+                let buf = &mut Cursor::new(Vec::new());
+                self.write_body(buf, options)?;
+                let buf = buf.get_ref();
+
+                cursor.write_string(stringify!($property))?;
+                cursor.write_u64::<LittleEndian>(buf.len() as u64)?;
+                $(
+                    impl_write_header_part!(self, cursor, $header_property);
```

Comment:
This seems slightly confusing to use, maybe make the macros so that the write implementation looks like so?
```rust
impl PropertyTrait for SetProperty {
#[inline]
            fn write<W: Write>(
                &self,
                cursor: &mut W,
                include_header: bool,
                options: &mut PropertyOptions,
            ) -> Result<(), Error> {
                write_prologue!(self, cursor, include_header);
                cursor.write_string(&self.property_type)?;
                write_epilogue!(self, cursor, include_header);
            }
}
```
and if there's no custom logic there would be a macro that just groups prologue+epilogue

Reply:
That won't compile because variables defined in the prologue are scoped to the prologue and not accessible elsewhere.